Ulid modernization

marcin-dardzinski opened the pull request:
Addressing ulid intrinsic comments from #505

Some more tests should probably be necessary...

marcin-dardzinski committed:
Make private ulid arrays immutable, use library provided unix epoch start

marcin-dardzinski committed:
Implement `IUtf8SpanFormattable`, remove `TryWriteStringify` methods

marcin-dardzinski committed:
Rewrite `Equals`/`CompareTo`/`GetHashCode` into span based methods

marcin-dardzinski committed:
Don't assume little endian in dates when constructing ulid

marcin-dardzinski committed:
Simplify guid interop, add more tests

Saancreed commented on src/Domain/LeanCode.DomainModels/Ulids/Ulid.cs:
```diff
@@ -576,179 +573,131 @@ public bool TryWriteBytes(Span<byte> destination)
-        span[24] = Base32Bytes[((randomness8 & 3) << 3) | ((randomness9 & 224) >> 5)];
-
-        return true;
+        return string.Create(BytesLength, this, (span, ulid) => ulid.TryFormat(span, out _, "", null));
```
Using `BytesLength` in this context suggests (and rightfully so) that this isn't actually length in _bytes_ but in base32 digits which all happen to be encodable with a single UTF8 code unit or a single UTF16 code unit each.

`LengthInChars` would be less incorrect, if we allow for overloaded use of term "character" :stuck_out_tongue:

Or [`LengthInTextElements`](https://learn.microsoft.com/en-us/dotnet/api/system.globalization.stringinfo.lengthintextelements?view=net-7.0) :see_no_evil:

Saancreed commented on src/Domain/LeanCode.DomainModels/Ulids/Ulid.cs:
```diff
@@ -759,92 +708,19 @@ public override bool Equals(object? obj)
-        {
-            return GetResult(this.randomness9, other.randomness9);
-        }
+    private static ReadOnlySpan<byte> AsSpan(in Ulid ulid)
```
`[MemoryImpl(MemoryImplOptions.AggressingInlining)]` on this? :see_no_evil:

jakubfijalkowski commented on src/Domain/LeanCode.DomainModels/Ulids/Ulid.cs:
```diff
@@ -367,23 +361,26 @@ public Ulid(Guid guid)
-            tmp = buf[6];
-            buf[6] = buf[7];
-            buf[7] = tmp;
+            buf[0..4].Reverse();
```
Can we get a comment here? The non-symmetry is quite mind-bending. ^^'

Saancreed reviewed (approved):
Two suggestions, otherwise LGTM.

marcin-dardzinski committed:
Ulid code review fixes

1. Aggressively inline Ulid.AsSpan()
2. Rename LengthInTextElements for Ulid string length
3. Explain endianness swapping in Guid conversions
